lib/pluginmanager/offline_plugin_packager.rb

```diff
@@ -0,0 +1,120 @@
+# encoding: utf-8
+require "pluginmanager/ui"
+require "pluginmanager/errors"
+require "bootstrap/environment"
+require "bootstrap/util/compress"
+require "paquet"
+require "stud/temporary"
+require "fileutils"
+
+module LogStash module PluginManager
+  class SpecificationHelpers
+    WILDCARD = "*"
+    WILDCARD_INTO_RE = ".*"
+
+    def self.find_by_name_with_wildcards(pattern)
+      re = transform_pattern_into_re(pattern)
+      ::Gem::Specification.find_all.select do |specification|
+        specification.name =~ re
+      end
+    end
+
+    def self.transform_pattern_into_re(pattern)
+      Regexp.new(/^#{pattern.gsub(WILDCARD, WILDCARD_INTO_RE)}$/)
+    end
+  end
+
+  class OfflinePluginPackager
+    LOGSTASH_DIR = "logstash"
+    DEPENDENCIES_DIR = ::File.join(LOGSTASH_DIR, "dependencies")
+
+    # To make sure we have the maximun compatibility
+    # we will ignore theses gems and they wont be included in the pack
+    IGNORE_GEMS_IN_PACK = %w(
+      logstash-core
+      logstash-core-queue-jruby
+      logstash-core-event-java
+      logstash-core-plugin-api
+      jar-dependencies
+    )
+
+    INVALID_PLUGINS_TO_EXPLICIT_PACK = IGNORE_GEMS_IN_PACK.collect { |name| Regexp.new(/^#{name}/) } + [
+      /mixin/
+    ]
+
+    def initialize(plugins_to_package, target)
+      @plugins_to_package = Array(plugins_to_package)
+      @target = target
+
+      validate_plugins!
+    end
+
+    def validate_plugins!
+      @plugins_to_package.each do |plugin_name|
+        if INVALID_PLUGINS_TO_EXPLICIT_PACK.any? { |invalid_name| plugin_name =~ invalid_name }
+          raise UnpackablePluginError, "Cannot explicitely pack `#{plugin_name}` for offline installation"
+        end
+      end
+    end
+
+    def generate_temporary_path
+      Stud::Temporary.pathname
+    end
+
+    def explicitely_declared_plugins_specs
+      @plugins_to_package.collect do |plugin_pattern|
+        specs = SpecificationHelpers.find_by_name_with_wildcards(plugin_pattern)
```

Comment:
what is the purpose of supporting wildcards here? I think it's safer to support only 1 to 1 declaration of plugins to be packed. this would mean that specs.size has to be 1. If it's 2 or more it should generate some exception because you can have multiple plugin specs for different versions (it's easy for `vendor/` to become littered with gem trash)